### PR TITLE
(BKR-1423) Fix SSH settings in nodesets

### PR DESCRIPTION
**Pull request by trevor-vaughan**
Allow SSH settings from nodesets to propagate properly into the host's
SSH configuration.

BKR-1423 #close

### CODE REVIEW

**Comment by trevor-vaughan** [comment]
@sbeaulie This is the correct fix instead of the PR in https://github.com/puppetlabs/beaker/pull/1633

File: lib/beaker/hypervisor/vagrant.rb [hunk 1]
```diff
@@ -201,7 +201,8 @@ def set_ssh_config host, user
 
         f.write(ssh_config)
         f.rewind
-        host['ssh'] = {:config => f.path()}
+
+        host['ssh'] = host['ssh'].merge(Net::SSH.configuration_for(host['ip'], f.path))
```
Comment (sbeaulie) [inline lib/beaker/hypervisor/vagrant.rb]: Thank for your contribution and testing!

I think this is an elegant solution. From what I see in the code these files are based on the command `vagrant ssh-config` and modified to use `host['ip']`, `host['user']`, and set `IdentitiesOnly no`. So this would replace the `host['ssh']['user']` if I'm understanding it well.

If that is the correct behavior, can we document that somewhere?

Lastly, this change removes the previous `:config` key,  my understanding is that it is needed for the rsync method, so let's keep it in order not to break that method.

```suggestion
        host['ssh'][:config] = f.path()
        host['ssh'] = host['ssh'].merge(Net::SSH.configuration_for(host['ip'], f.path))
```

**Comment by trevor-vaughan** [comment]
@sbeaulie The issue is that if the `:config` item is populated, literally everything else is overridden which was the crux of the original problem. We may need an auxillary data structure for `rsync`.

**Comment by sbeaulie** [comment]
I've traced the code and the only place it seems to be used is with the rsynch method in beaker so I don't see where it would conflict - can you shed light on where it would replace everything? I can test my suggestion with the `pupmod-simp-acpid` project you suggested.

**Comment by trevor-vaughan** [comment]
The `host['ssh']` hash is passed directly into the Net::SSH connection configuration. If `host['ssh'][:config]` is populated, Net::SSH will pick it up and run with it.

We almost need a `host['ssh'][:cli_config]` or something.

**Comment by trevor-vaughan** [comment]
Testing a solution now.

**Comment by sbeaulie** [comment]
This is the code where it is being used: https://github.com/puppetlabs/beaker/blob/master/lib/beaker/host.rb#L550-L552

We could change the name in both places. But I still don't see where NET::SSH would pick the `host['ssh'][:config]` up.

I tested with having both `:config` and the merge and it passed in `pupmod-simp-acpid`. The connection attempts look like this:

```
Attempting ssh connection to 10.X.X.X, user: root, opts: {:config=>"/var/folders/yb/y8xggbwd5l17w64zjz174h4c0000gn/T/el620200318-3752-y2mnkr", :verify_host_key=>false, :auth_methods=>["none", "publickey", "keyboard-interactive"], :port=>2222, :forward_agent=>true, :keys=>["XX/insecure_private_key"], :user_known_hosts_file=>"/dev/null", :keepalive=>true, :host_name=>"127.0.0.1", :user=>"root", :keys_only=>true, :logger=>...}
```

**Comment by sbeaulie** [comment]
Can I ask you to test with my suggested change and give me the details on the failure you get?

Thanks!

**Comment by trevor-vaughan** [comment]
It will work, because it's ignoring everything else.

See the `:config` option in Net::SSH itself https://www.rubydoc.info/github/net-ssh/net-ssh/Net/SSH

This basically seems like it's fixing the original issue, but it's not because all of the other options in the hash are now ignored.

**Comment by trevor-vaughan** [comment]
If you try sticking something that Net::SSH can't handle into the Hash, you can see it fail.

**Comment by trevor-vaughan** [comment]
Ok, this commit, when combined with https://github.com/puppetlabs/beaker/pull/1634 should do the right thing in all cases (I think)

**Comment by sbeaulie** [comment]
ok I'm following now 😄 From the Net SSH config class documents:

> Note that you will never need to use this class directly–you can control whether the OpenSSH configuration files are read by passing the :config option to Net::SSH.start. (They are, by default.)

In beaker we run start like so [Net::SSH.start(host, user, ssh_opts) ](https://github.com/puppetlabs/beaker/blob/master/lib/beaker/ssh_connection.rb#L71) where `ssh_opts[:config]` can be set as follows 
> :config => set to true to load the default OpenSSH config files (~/.ssh/config, /etc/ssh_config), or to false to not load them, or to a file-name (or array of file-names) to load those specific configuration files. Defaults to true.

If we look at the source code for Net:SSH version 5.0 used in beaker you are correct that the ssh_opts hash will return an ArgumentError for any non valid options:
[github](https://github.com/net-ssh/net-ssh/blob/v5.0.0/lib/net/ssh.rb#L214)

The only thing in the start method that is different from the experience you describe is that it seems like it would load the file from `:config` (if its a file path) and then merge the other options on top, so the passed hash should take precedence, and not the `:config`:
https://github.com/net-ssh/net-ssh/blob/v5.0.0/lib/net/ssh.rb#L221

In conclusion, I'm happy with this fix, `:vagrant_ssh_config` should not be put in the 'ssh' hash, so putting it within the host is fine. Is there any documentation that would need updated for this new field? Since we do not make that variable avalable to users, and its only used for internal logic between beaker-vagrant and beaker I think it is fine without.

One thing left to fix is the Unit tests. See the travis results: https://travis-ci.com/github/puppetlabs/beaker-vagrant/jobs/299555431
I think the first failed test's expect can be changed to the new `:vagrant_ssh_config` and the second failed test's expect is probably not needed.

File: lib/beaker/hypervisor/vagrant.rb [hunk 2]
```diff
@@ -201,7 +201,9 @@ def set_ssh_config host, user
 
         f.write(ssh_config)
         f.rewind
-        host['ssh'] = {:config => f.path()}
+
+        host[:vagrant_ssh_config] = f.path
```
Comment (sbeaulie) [inline lib/beaker/hypervisor/vagrant.rb]: needs the respective unit test fixed.

Comment (trevor-vaughan) [reply lib/beaker/hypervisor/vagrant.rb]: Working on that now
